libp2p/Session.cpp

```diff
@@ -300,13 +303,23 @@ void Session::disconnect(DisconnectReason _reason)
 {
     clog(VerbosityTrace, "p2pcap") << "Disconnecting (our reason: " << reasonOf(_reason) << ") from " << m_logSuffix;
 
-    if (m_socket->ref().is_open())
+    if (!m_dropped)
     {
         RLPStream s;
         prep(s, DisconnectPacket, 1) << (int)_reason;
         sealAndSend(s);
+        auto disconnectTimer = m_server->createTimer();
```

Comment:
This code will still close the socket immediately (`Session::drop` closes the socket and you're calling it in both your timer handler and outside the handler if it hasn't already been called i.e. !m_dropped). Since what we want to do is delay the socket closure for 2 seconds, you should move the code which closes the socket out of Session::drop and into the timer handler and also not call drop inside of the timer handler.

Reply:
@halfalicious  
Thanks for the reviews!
Okay so I have to copy required logic of `drop` into `disconnect` and never call `drop` from inside of `disconnect` and its handler. Am I right??

Reply:
You should move the socket closure code out of drop and into the handler. Don’t call drop inside of the handler but still call it in your else statement in disconnect (since we want to prevent further session operations after we have sent the disconnect, you can see that we check for m_drop in various places in session and return early)

Reply:
@halfalicious 
So pushed changes according to your suggested changes. But was curious what will happen to code thats calling `drop()` (apart from `disconnect`) separately??

Reply:
This still doesn't look right:
1. Changing `drop()` now breaks all the cases where it is called from other methods of `Session`. I guess those are reasonable cases where we wanted to just disconnect immediately without sending `Disconnect` packet (with current changes we won't disconnect)

2. Calling `disconnect` now doesn't ever set `m_dropped` to `true`. @halfalicious Maybe you meant calling current form of `drop` from inside `if (!m_dropped)` clause?

Reply:
@gumb0 @halfalicious 
So what we can do is create a overloaded copy of drop() since functionality of drop is now slightly different in 2 cases. (one where normal code calls it. two where disconnect calls it). We can place second argument of drop() as bool or number of seconds(which will be 2). 

Will this work out??

Reply:
@gumb0 You're right we need to also call drop() inside of the !m_dropped clause. 

I don't think this new form of drop breaks anything though...even if drop doesn't close the socket all shared_ptr Session instances will still go out of scope and therefore Session will be garbage collected by Host and the socket will be closed when the dtor is invoked. Note that functions in which we pass new shared_ptr instances (shared_from_this) to handlers check `if (m_dropped)` and then prematurely terminate before creating new handlers (e.g. `doRead`).  In fact, after taking a closer look at the Session code, I don't think we need to explicitly close the socket at all in any place other than the dtor (unless there's some benefit to closing the socket sooner rather than later that I'm not aware of)? If we just call drop() without closing the socket the Session instance will be destroyed once all queued handlers which contain a shared_ptr instance are executed and the Host garbage collects the Session.

So, to summarize:
* I think we should call drop() in both the if (!m_dropped) and else cases...in fact, we can simplify this by removing the "else" case and always calling drop() after  the `if (!m_dropped)` code block since `drop `returns early if `m_dropped` is true.
* I think we can remove all cases where we close the socket except for in the dtor. 

@gumb0 : Thoughts?

Reply:
>  I don't think we need to explicitly close the socket at all in any place other than the dtor (unless there's some benefit to closing the socket sooner rather than later that I'm not aware of)? 

@halfalicious The benefit I see is that we immediately stop answering any further requests on this connection, not waiting for all handlers to terminate, so maybe it's more efficient...

Also, what if some malicious peer keeps DOSing us with some invalid requests on this connection - it will be kept in the handler queue (becouse the new data keeps coming) and never disconnected, it seems so.

Reply:
@halfalicious pointed out to me:

> I think that we will still stop reading packets from a peer if we call drop without closing the socket since we check for m_dropped at the beginning of Session::doRead before proceeding with the read: https://github.com/ethereum/aleth/blob/d8f4498f5b2bfa218c40a0cc760b6a30a4d946c1/libp2p/Session.cpp#L318-L322

I think I agree with this reasoning.

Reply:
One corner case to check: spec requires that we disconnet immediately upon receiving Disconnect packet ourselves. My experiment confirms that this looks to be working correctly with current changes in this PR:
```
TRACE 08-08 11:54:24 p2p  p2pcap Disconnect from (##a8768590…@51.141.78.53:30303)
DEBUG 08-08 11:54:24 p2p  p2pcap Disconnect (reason: Peer had too many connections.) from (##a8768590…@51.141.78.53:30303)
DEBUG 08-08 11:54:24 p2p  net    Closing peer session with (##a8768590…@51.141.78.53:30303)
```

Reply:
> 
> 
> One corner case to check: spec requires that we disconnet immediately upon receiving Disconnect packet ourselves. My experiment confirms that this looks to be working correctly with current changes in this PR:
> 
> ```
> TRACE 08-08 11:54:24 p2p  p2pcap Disconnect from (##a8768590…@51.141.78.53:30303)
> DEBUG 08-08 11:54:24 p2p  p2pcap Disconnect (reason: Peer had too many connections.) from (##a8768590…@51.141.78.53:30303)
> DEBUG 08-08 11:54:24 p2p  net    Closing peer session with (##a8768590…@51.141.78.53:30303)
> ```

Confirmed by looking at the code - when we receive a disconnect packet we call `drop ` immediately in `Session::interpretP2pPacket`: https://github.com/ethereum/aleth/blob/78355a28da55f1c0e952e3cf1a17891cea981cb4/libp2p/Session.cpp#L142-L154